public/static/sdk/jsmr-0.1.js

```diff
@@ -145,22 +196,113 @@ var JsMr = Class.extend({
         );
     },
 
+    getData: function(task) {
+
+
+        var output = "";
+
+        var context = {
+            write: function (key, value) {
+                output = output.concat(key, ":", value, "\t");
+            },
+            state: {}
+        }
+
+        var _this = this;
+        var credentials = {accessKeyId: task.access_key, secretAccessKey: task.secret_key};
+        AWS.config.update(credentials);
+        //AWS.config.region = 'us-west-2';
+        var s3 = new AWS.S3();
+
+        //Fetch state from AWS
+         var job_id = task.job_id.toString();
+         var state_params = {
+             Bucket: task.bucket_name,
+             Key: job_id.concat('_state'),
+         }
+
+         s3.getObject(state_params, function (err, data) {
+         if (err){
+	     console.log("Error in fetching state");
+             console.log(err, err.stack); // an error occurred
+	 }
+         else {
+             // successful response
+             this.current_state = JSON.parse( data.Body.toString());
+             context.state = this.current_state;
+	     console.log("Fetched state is: ");
+	     console.log(this.current_state);
+         }
+         });
+	
+        //Fetch data from AWS
+        var params = {
+            Bucket: task.bucket_name,
+            Key: task.object_key,
+            Range: 'bytes='.concat(task.start_index, '-', task.end_index)
+        };
+
+        var data_chunk;
+        s3.getObject(params, function (err, data) {
+            if (err){
+		console.log("Error in get Object");
+                console.log(err, err.stack); // an error occurred
+	    }
+            else {
+                // successful response
+		console.log("Successfull get object");
+                data_chunk = data.Body.toString();
+                var split_data = data_chunk.split("\n");
+
+                runMap().setup(context);
+                //run code on each line of data
+                for (var i = 0; i < split_data.length; i++) {
```

Comment:
At this point the state might or might not be loaded because it's an Async call. They need to be chained in order to guarantee that they are both loaded. I mean:
```js
s3.getObject(the_state..., function(err, data) {
    if (err) ...
    else {
        s3.getObject(the_data..., function(err_2, data_2) {
            if (err) ...
            else {
                // now iterate and do other things
            }
        });
    }
});
```